src/CapabilityService.WebApi/Application/CapabilityApplicationService.cs

```diff
@@ -73,11 +83,12 @@ public async Task AddContext(Guid capabilityId, string contextName)
             {
                 throw new CapabilityDoesNotExistException();
             }
-            
+
             capability.AddContext(contextName);
         }
 
-        public async Task UpdateContext(Guid capabilityId, Guid contextId, string awsAccountId, string awsRoleArn, string awsRoleEmail)
+        public async Task UpdateContext(Guid capabilityId, Guid contextId, string awsAccountId, string awsRoleArn,
+            string awsRoleEmail)
         {
             var capability = await _capabilityRepository.Get(capabilityId);
             if (capability == null)

```

Comment:
I think we should throw an argument exception instead and in general avoid custom exceptions for cases where we cant load data based on the input parameters.

Reply:
I would like to move the exception into the CapabilityRepository, since all consumers of the capabilityRepository.Get makes the same if check.
To me a argument exception should only tell you something about the argument given. Like null is not accepter, the value is out of range or that you can't have a exclamation mark in a name.

Reply:
I guess it all comes down to how many exception classes we want and personally I have been schooled to avoid creating custom exceptions whenever its not absolutely necessary. However moving it into the actual Get method would probably be a bad idea, I understand the 80/20 rule might make it appealing, but as a side effect we would make every "null search" throw an exception, which costs more performance wise then returning null... the better solution would probably be to remove the null check and exception all together and do something like this: capability?.UpdateContext and leave it for the compile to sort out the rest :)

Reply:
I think exceptions sucks. Unfortunately I believe they are our best vessel of alternate control flows unless we take a more radical approach to altering the control like options/either or action results.
Custom exceptions empowers us to be very specific about why we halted a execution which to me makes it much easier to react to.
In this case throwing the `CapabilityDoesNotExistException` allows us to catch it a API controller and translate it to a 404 status code.
I think this task would have been much harder had we thrown a argument exception.

Performance will likely be something we debate more in the future. I will chose many other things over performance. In this case I feel that the value of gaining access to the control flow is worth more than the cost of throwing exceptions.

Reply:
With great power comes great responsibilty.. ;) and I think throwing a custom exception because we want to inform someone of the fact that we could not do an update because an input argument did not succesfully resolve to a entity leads us down a path where we will have alot of custom exceptions. :) 

Having said that I am more concerned that you will choose many other things over performance. Performance is the reason why MS is introducing things like ValueType and IAsyncEnumerable to the framework, because milliseconds matter in scalable microservice architectures. Often more then the quality of the code granted that the popular opinion is that a MS should be written in a few hundred lines of code :)